docker-java-core/src/main/java/com/github/dockerjava/core/DefaultDockerCmdExecFactory.java

```diff
@@ -0,0 +1,154 @@
+            );
+        }
+
+        DefaultWebTarget(
+            ImmutableList<String> path,
+            SetMultimap<String, String> queryParams
+        ) {
+            this.path = path;
+            this.queryParams = queryParams;
+        }
+
+        @Override
+        public String toString() {
+            return String.format("DefaultWebTarget{path=%s, queryParams=%s}", path, queryParams);
+        }
+
+        @Override
+        public InvocationBuilder request() {
+            String resource = StringUtils.join(path, "/");
+
+            if (!resource.startsWith("/")) {
+                resource = "/" + resource;
+            }
+
+            if (!queryParams.isEmpty()) {
+                Escaper urlFormParameterEscaper = UrlEscapers.urlFormParameterEscaper();
+                resource = queryParams.asMap().entrySet().stream()
+                    .flatMap(entry -> {
+                        return entry.getValue().stream().map(s -> {
+                            return entry.getKey() + "=" + urlFormParameterEscaper.escape(s);
+                        });
+                    })
+                    .collect(Collectors.joining("&", resource + "?", ""));
+            }
+
+            return new DefaultInvocationBuilder(
+                dockerHttpClient, objectMapper, resource
+            );
+        }
+
+        @Override
+        public DefaultWebTarget path(String... components) {
+            ImmutableList<String> newPath = ImmutableList.<String>builder()
+                .addAll(path)
+                .add(components)
+                .build();
+            return new DefaultWebTarget(newPath, queryParams);
+        }
+
+        @Override
+        public DefaultWebTarget resolveTemplate(String name, Object value) {
+            ImmutableList.Builder<String> newPath = ImmutableList.builder();
+            for (String component : path) {
+                component = component.replaceAll(
+                    "\\{" + name + "\\}",
+                    UrlEscapers.urlPathSegmentEscaper().escape(value.toString())
+                );
+                newPath.add(component);
+            }
+
+            return new DefaultWebTarget(newPath.build(), queryParams);
+        }
+
+        @Override
+        public DefaultWebTarget queryParam(String name, Object value) {
+            if (value == null) {
+                return this;
+            }
+
+            SetMultimap<String, String> newQueryParams = HashMultimap.create(queryParams);
+            newQueryParams.put(name, value.toString());
+
+            return new DefaultWebTarget(path, newQueryParams);
+        }
+
+        @Override
+        public DefaultWebTarget queryParamsSet(String name, Set<?> values) {
+            SetMultimap<String, String> newQueryParams = HashMultimap.create(queryParams);
+            newQueryParams.replaceValues(name, values.stream().filter(Objects::nonNull).map(Object::toString).collect(Collectors.toSet()));
```

Comment:
let's extract second param, `nonNullValues`

Reply:
WDYT about doing a follow up? This class is copied almost as-is and I would prefer to keep it so to make it easier to review the actual change